src/documentation/programming-guide.md

```diff
@@ -1013,8 +1014,186 @@ odds = results.odd
 tens = results[None]  # the undeclared main output
 ```
 
-<a name="io"></a>
-<a name="running"></a>
+## <a name="io"></a>Pipeline I/O
+
+When you create a pipeline, you often need to read data from some external source, such as a file in external data sink or a database. Likewise, you may want your pipeline to output its result data to a similar external data sink. Beam provides read and write transforms for a number of common data storage types. If you want your pipeline to read from or write to a data storage format that isn't supported by the built-in transforms, you can implement your own read and write transforms.
+
+> A guide that covers how to implement your own Beam IO transforms is in progress ([BEAM-1025](https://issues.apache.org/jira/browse/BEAM-1025)).
+
+### Reading Input Data
```

Comment:
I would drop the "two ways to..." and just recommend using the transform way only. We do not believe it is good practice to expose `Source`/`Sink` objects -- they are for transform authors not pipeline authors.